File: src/test/java/com/mpe85/grampa/util/stack/impl/RestorableStackTest.java
package com.mpe85.grampa.util.stack.impl;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

import com.mpe85.grampa.util.stack.RestorableStack;

public class RestorableStackTest {
	
	@Test
	public void test_push_valid() {
		final RestorableStack<Number> stack1 = new LinkedListRestorableStack<>();
		final RestorableStack<Number> stack2 = new LinkedListRestorableStack<>();
		
		stack1.push(1);
		stack1.push(2);
		stack1.push(3);
		
		stack2.push(0, 3);
		stack2.push(1, 2);
		stack2.push(2, 1);
		
		assertEquals(3, stack1.size());
		assertEquals(3, stack2.size());
		assertEquals(stack1, stack2);
	}
	
	@Test
	public void test_peek_valid() {
		final RestorableStack<Number> stack = new LinkedListRestorableStack<>();
		
		stack.push(1);
		stack.push(2);
		stack.push(3);
		
		assertEquals(3, stack.size());
		assertEquals(3, stack.peek());
		assertEquals(3, stack.peek(0));
		assertEquals(2, stack.peek(1));
		assertEquals(1, stack.peek(2));
		assertEquals(3, stack.size());
	}
	
	@Test
	public void test_peekAs_valid() {
		final RestorableStack<Number> stack = new LinkedListRestorableStack<>();
		
		stack.push(1);
		stack.push(2);
		stack.push(3.3f);
		
		assertEquals(3, stack.size());
		assertEquals(3.3f, stack.peekAs(Float.class).floatValue());
		assertEquals(1, stack.peekAs(2, Integer.class).intValue());
		assertEquals(3, stack.size());
	}
	
	@Test
	public void test_pop_valid() {
		final RestorableStack<Number> stack = new LinkedListRestorableStack<>();
		
		stack.push(1);
		stack.push(2);
		stack.push(3);
		
		assertEquals(3, stack.size());
		assertEquals(3, stack.pop());
		assertEquals(2, stack.size());
		assertEquals(1, stack.pop(1));
		assertEquals(1, stack.size());
		assertEquals(2, stack.pop(0));
		assertEquals(0, stack.size());
	}
	
	@Test
	public void test_popAS_valid() {
		final RestorableStack<Number> stack = new LinkedListRestorableStack<>();
		
		stack.push(1);
		stack.push(2.2d);
		stack.push(3);
		
		assertEquals(3, stack.size());
		assertEquals(2.2d, stack.popAs(1, Double.class).doubleValue());
		assertEquals(2, stack.size());
		assertEquals(3, stack.popAs(Integer.class).intValue());
	}
	
	@Test
	public void test_poke_valid() {
		final RestorableStack<Number> stack = new LinkedListRestorableStack<>();
		
		stack.push(1);
		stack.push(2);
		stack.push(3);
		
		assertEquals(3, stack.size());
		assertEquals(3, stack.poke(4));
		assertEquals(4, stack.peek());
		assertEquals(3, stack.size());
		assertEquals(2, stack.poke(1, 5));
		assertEquals(5, stack.peek(1));
		assertEquals(3, stack.size());
	}
	
	@Test
	public void test_dup_valid() {
		final RestorableStack<Number> stack = new LinkedListRestorableStack<>();
		
		stack.push(1);
		stack.dup();
		stack.dup();
		
		assertEquals(3, stack.size());
		assertEquals(1, stack.peek());
		assertEquals(1, stack.peek(1));
		assertEquals(1, stack.peek(2));
	}
	
	@Test
	public void test_swap_valid() {
		final RestorableStack<Number> stack = new LinkedListRestorableStack<>();
		
		stack.push(1);
		stack.push(2);
		stack.push(3);
		stack.swap();
		
		assertEquals(3, stack.size());
		assertEquals(2, stack.peek());
		assertEquals(3, stack.peek(1));
		assertEquals(1, stack.peek(2));
	}
	
	@Test
	public void test_restoreSnapshot_valid() {
		final RestorableStack<Number> stack = new LinkedListRestorableStack<>();
		
		stack.push(1);
		stack.takeSnapshot();
		stack.push(2);
		stack.push(3);
		stack.takeSnapshot();
		
		stack.pop(1);
		
		assertEquals(2, stack.size());
		assertEquals(2, stack.getSnapshotCount());
		
		stack.restoreSnapshot();
		
		assertEquals(3, stack.size());
		assertEquals(3, stack.peek());
		assertEquals(2, stack.peek(1));
		assertEquals(1, stack.peek(2));
		assertEquals(1, stack.getSnapshotCount());
		
		stack.restoreSnapshot();
		
		assertEquals(1, stack.size());
		assertEquals(1, stack.peek());
		assertEquals(0, stack.getSnapshotCount());
	}
	
	@Test
	public void test_discardSnapshot_valid() {
		final RestorableStack<Number> stack = new LinkedListRestorableStack<>();
		
		stack.push(1);
		stack.takeSnapshot();
		stack.push(2);
		stack.takeSnapshot();
		
		assertEquals(2, stack.getSnapshotCount());
		
		stack.discardSnapshot();
		
		assertEquals(1, stack.getSnapshotCount());
		
		stack.discardSnapshot();
		
		assertEquals(0, stack.getSnapshotCount());
	}
	
	@Test
	public void test_clearAllSnapshots_valid() {
		final RestorableStack<Number> stack = new LinkedListRestorableStack<>();
		
		stack.push(1);
		stack.takeSnapshot();
		stack.push(2);
		stack.takeSnapshot();
		
		assertEquals(2, stack.getSnapshotCount());
		
		stack.clearAllSnapshots();
		
		assertEquals(0, stack.getSnapshotCount());
	}
	
}

File: src/test/kotlin/com/mpe85/grampa/util/stack/impl/RestorableStackTest.kt
package com.mpe85.grampa.util.stack.impl

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class RestorableStackTest {

  @Test
  fun test_push_valid() {
    val stack1 = LinkedListRestorableStack<Number>()
    val stack2 = LinkedListRestorableStack<Number>()

    stack1.push(1)
    stack1.push(2)
    stack1.push(3)

    stack2.push(0, 3)
    stack2.push(1, 2)
    stack2.push(2, 1)

    Assertions.assertEquals(3, stack1.size)
    Assertions.assertEquals(3, stack2.size)
    Assertions.assertEquals(stack1, stack2)
  }

  @Test
  fun test_peek_valid() {
    val stack = LinkedListRestorableStack<Number>()

    stack.push(1)
    stack.push(2)
    stack.push(3)

    Assertions.assertEquals(3, stack.size)
    Assertions.assertEquals(3, stack.peek())
    Assertions.assertEquals(3, stack.peek(0))
    Assertions.assertEquals(2, stack.peek(1))
    Assertions.assertEquals(1, stack.peek(2))
    Assertions.assertEquals(3, stack.size)
  }

  @Test
  fun test_peekAs_valid() {
    val stack = LinkedListRestorableStack<Number>()

    stack.push(1)
    stack.push(2)
    stack.push(3.3f)

    Assertions.assertEquals(3, stack.size)
    Assertions.assertEquals(3.3f, stack.peekAs(Float::class.javaObjectType).toFloat())
    Assertions.assertEquals(1, stack.peekAs(2, Int::class.javaObjectType).toInt())
    Assertions.assertEquals(3, stack.size)
  }

  @Test
  fun test_pop_valid() {
    val stack = LinkedListRestorableStack<Number>()

    stack.push(1)
    stack.push(2)
    stack.push(3)

    Assertions.assertEquals(3, stack.size)
    Assertions.assertEquals(3, stack.pop())
    Assertions.assertEquals(2, stack.size)
    Assertions.assertEquals(1, stack.pop(1))
    Assertions.assertEquals(1, stack.size)
    Assertions.assertEquals(2, stack.pop(0))
    Assertions.assertEquals(0, stack.size)
  }

  @Test
  fun test_popAS_valid() {
    val stack = LinkedListRestorableStack<Number>()

    stack.push(1)
    stack.push(2.2)
    stack.push(3)

    Assertions.assertEquals(3, stack.size)
    Assertions.assertEquals(2.2, stack.popAs(1, Double::class.javaObjectType).toDouble())
    Assertions.assertEquals(2, stack.size)
    Assertions.assertEquals(3, stack.popAs(Int::class.javaObjectType).toInt())
  }

  @Test
  fun test_poke_valid() {
    val stack = LinkedListRestorableStack<Number>()

    stack.push(1)
    stack.push(2)
    stack.push(3)

    Assertions.assertEquals(3, stack.size)
    Assertions.assertEquals(3, stack.poke(4))
    Assertions.assertEquals(4, stack.peek())
    Assertions.assertEquals(3, stack.size)
    Assertions.assertEquals(2, stack.poke(1, 5))
    Assertions.assertEquals(5, stack.peek(1))
    Assertions.assertEquals(3, stack.size)
  }

  @Test
  fun test_dup_valid() {
    val stack = LinkedListRestorableStack<Number>()

    stack.push(1)
    stack.dup()
    stack.dup()

    Assertions.assertEquals(3, stack.size)
    Assertions.assertEquals(1, stack.peek())
    Assertions.assertEquals(1, stack.peek(1))
    Assertions.assertEquals(1, stack.peek(2))
  }

  @Test
  fun test_swap_valid() {
    val stack = LinkedListRestorableStack<Number>()

    stack.push(1)
    stack.push(2)
    stack.push(3)
    stack.swap()

    Assertions.assertEquals(3, stack.size)
    Assertions.assertEquals(2, stack.peek())
    Assertions.assertEquals(3, stack.peek(1))
    Assertions.assertEquals(1, stack.peek(2))
  }

  @Test
  fun test_restoreSnapshot_valid() {
    val stack = LinkedListRestorableStack<Number>()

    stack.push(1)
    stack.takeSnapshot()
    stack.push(2)
    stack.push(3)
    stack.takeSnapshot()
    stack.pop(1)

    Assertions.assertEquals(2, stack.size)
    Assertions.assertEquals(2, stack.snapshotCount)

    stack.restoreSnapshot()

    Assertions.assertEquals(3, stack.size)
    Assertions.assertEquals(3, stack.peek())
    Assertions.assertEquals(2, stack.peek(1))
    Assertions.assertEquals(1, stack.peek(2))
    Assertions.assertEquals(1, stack.snapshotCount)

    stack.restoreSnapshot()

    Assertions.assertEquals(1, stack.size)
    Assertions.assertEquals(1, stack.peek())
    Assertions.assertEquals(0, stack.snapshotCount)
  }

  @Test
  fun test_discardSnapshot_valid() {
    val stack = LinkedListRestorableStack<Number>()

    stack.push(1)
    stack.takeSnapshot()
    stack.push(2)
    stack.takeSnapshot()

    Assertions.assertEquals(2, stack.snapshotCount)

    stack.discardSnapshot()

    Assertions.assertEquals(1, stack.snapshotCount)

    stack.discardSnapshot()

    Assertions.assertEquals(0, stack.snapshotCount)
  }

  @Test
  fun test_clearAllSnapshots_valid() {
    val stack = LinkedListRestorableStack<Number>()

    stack.push(1)
    stack.takeSnapshot()
    stack.push(2)
    stack.takeSnapshot()

    Assertions.assertEquals(2, stack.snapshotCount)

    stack.clearAllSnapshots()

    Assertions.assertEquals(0, stack.snapshotCount)
  }
}